io/eolearn/io/extra/meteoblue.py

```diff
@@ -76,67 +78,68 @@ def __init__(
         self.time_difference = time_difference
 
     @staticmethod
-    def _get_modified_eopatch(eopatch: Optional[EOPatch], bbox: Optional[BBox]) -> EOPatch:
+    def _get_modified_eopatch(eopatch: Optional[EOPatch], bbox: Optional[BBox]) -> Tuple[BBox, EOPatch]:
         if bbox is not None:
             if eopatch is None:
                 eopatch = EOPatch(bbox=bbox)
             elif eopatch.bbox is None:
                 eopatch.bbox = bbox
             elif eopatch.bbox != bbox:
                 raise ValueError("Provided eopatch.bbox and bbox are not the same")
-            return eopatch
+            return bbox, eopatch
 
         if eopatch is None or eopatch.bbox is None:
             raise ValueError("Bounding box is not provided")
-        return eopatch
+        return eopatch.bbox, eopatch
 
-    def _prepare_time_intervals(self, eopatch, time_interval):
+    def _prepare_time_intervals(self, eopatch: EOPatch, time_interval: Optional[RawTimeIntervalType]) -> List[str]:
         """Prepare a list of time intervals for which data will be collected from meteoblue services"""
         if not eopatch.timestamps and not time_interval:
             raise ValueError(
                 "Time interval should either be defined with `eopatch.timestamps` or `time_interval` parameter"
             )
 
         if time_interval:
-            start_time, end_time = serialize_time(parse_time_interval(time_interval))
-            return [f"{start_time}/{end_time}"]
+            serialized_start_time, serialized_end_time = serialize_time(parse_time_interval(time_interval))
+            return [f"{serialized_start_time}/{serialized_end_time}"]
 
         timestamps = eopatch.timestamps
-        time_intervals = []
+        time_intervals: List[str] = []
         for timestamp in timestamps:
             start_time = timestamp - self.time_difference
             end_time = timestamp + self.time_difference
 
-            start_time, end_time = serialize_time((start_time, end_time))
-            time_interval = f"{start_time}/{end_time}"
+            serizalized_start_time, serizalized_end_time = serialize_time((start_time, end_time))
 
-            time_intervals.append(time_interval)
+            time_intervals.append(f"{serizalized_start_time}/{serizalized_end_time}")
 
         return time_intervals
 
-    def _get_data(self, query):
+    @abstractmethod
+    def _get_data(self, query: dict) -> Tuple[Any, List[dt.datetime]]:
         """It should return an output feature object and a list of timestamps"""
-        raise NotImplementedError
```

Comment:
the `NotImplementedError` way is not detected by any linter/type-checker, but when you use abstract methods it raises flags if a subclass does not implement the methods. Also you cannot use such a class in python until you override the abstract methods. Much safer :D

But in some cases this is not the desired behavior, mostly because we have some methods that certain child classes avoid. That is bad code design on our part, but that is out of this scope.